libraries/AP_HAL_Linux/Perf.cpp

```diff
@@ -15,173 +15,217 @@
-                            perf_elapsed->base.name);
+
+    if (perf.start == 0) {
+        hal.console->printf("perf_begin() called before begin() on perf_counter_t(%s)\n",
+                            perf.name);
         return;
     }
 
-    struct timespec ts;
-    clock_gettime(CLOCK_MONOTONIC, &ts);
-    const uint64_t elapsed = timespec_to_nsec(&ts) - perf_elapsed->start;
+    _update_count++;
+
+    const uint64_t elapsed = now_nsec() - perf.start;
+    perf.count++;
+    perf.total += elapsed;
 
-    perf_elapsed->count++;
-    perf_elapsed->total += elapsed;
+    if (perf.min > elapsed) {
+        perf.min = elapsed;
+    }
 
-    if (perf_elapsed->least > elapsed) {
-        perf_elapsed->least =  elapsed;
+    if (perf.max < elapsed) {
+        perf.max = elapsed;
     }
 
-    if (perf_elapsed->most < elapsed) {
-        perf_elapsed->most = elapsed;
+    if (!perf.count) {
+        /* when count wraps, just start treating new events with the higher
+         * weight instead of reseting the counter */
+        perf.count = 1;
     }
 
     /*
-     * Maintain mean and variance of interval in nanoseconds
-     * Knuth/Welford recursive mean and variance of update intervals (via Wikipedia)
+     * Maintain avg and variance of interval in nanoseconds
+     * Knuth/Welford recursive avg and variance of update intervals (via Wikipedia)
      * Same implementation of PX4.
      */
-    const double delta_intvl = elapsed - perf_elapsed->mean;
-    perf_elapsed->mean += (delta_intvl / perf_elapsed->count);
-    perf_elapsed->m2 += (delta_intvl * (elapsed - perf_elapsed->mean));
+    const double delta_intvl = elapsed - perf.avg;
+    perf.avg += (delta_intvl / perf.count);
+    perf.m2 += (delta_intvl * (elapsed - perf.avg));
+    perf.start = 0;
 
-    perf_elapsed->start = 0;
+    perf.lttng.end(perf.name);
 }
 
-void Util::perf_count(perf_counter_t perf)
+void Perf::count(Util::perf_counter_t pc)
 {
-    struct perf_counter_count_t *perf_counter = (struct perf_counter_count_t *)perf;
+    uintptr_t idx = (uintptr_t)pc;
 
-    if (perf_counter == NULL) {
+    if (idx >= _perf_counters.size()) {
         return;
     }
 
-    if (perf_counter->base.type != PC_COUNT) {
-        hal.console->printf("perf_count() called over a perf_counter_t(%s) "
-                            "that is not of the PC_COUNT type.\n",
-                            perf_counter->base.name);
+    Perf_Counter &perf = _perf_counters[idx];
+    if (perf.type != Util::PC_COUNT) {
+        hal.console->printf("perf_begin() called on perf_counter_t(%s) that"
+                            " is not of PC_COUNT type.\n",
+                            perf.name);
         return;
     }
 
-    perf_counter->count++;
+    _update_count++;
+    perf.count++;
+
+    perf.lttng.count(perf.name, perf.count);
 }
 
-#endif
+Util::perf_counter_t Perf::add(Util::perf_counter_type type, const char *name)
+{
+    if (type != Util::PC_COUNT && type != Util::PC_ELAPSED) {
```

Comment:
previously it was implemented in a switch/case and we had this:

```
-    default:
-    case PC_INTERVAL: {
-        /*
-         * Not implemented now because it is not used even on PX4 specific
-         * code and by looking at PX4 implementation without perf_reset()
-         * the average is broken.
-         */
```

I'm retaining this behavior because we need to check if it's one of those we implement, not one of those we don't. I'm rewording the comment a little bit to make that clear.